File: lang/nodejs/test/fun/vac/data_structures/DoublyLinkedListTest.js
"use strict"

const DoublyLinkedList = require("fun/vac/data_structures/DoublyLinkedList")
const TestRunner       = require("fun/vac/util/TestRunner")

const testLinkedList = () => {
    let size = 8192

    const list = new DoublyLinkedList.DoublyLinkedList()

    for (let i = 1; i <= size; i++) {
        list.append(i)
    }

    if (list.size() !== size) {
        return false
    }

    for (let i = 0; i < size; i++) {
        if (list.get(i) !== i + 1) {
            return false
        }
    }

    for (let i = 0; i < size; i++) {
        list.set(i, size - i)
    }

    for (let i = 0; i < size; i++) {
        if (list.get(i) !== size - i) {
            return false
        }
    }

    let x = 0, y = 0

    for (let i = 0, j = size - 1; i < j; i++, j--) {
        x = list.get(i)
        y = list.get(j)

        list.remove(i)
        list.insert(i, y)
        list.remove(j)
        list.insert(j, x)
    }

    for (let i = 0; i < size; i++) {
        if (list.get(i) !== i + 1) {
            return false
        }
    }

    for (let i = size; i >= 1; i--) {
        if (list.back() !== i) {
            return false
        } else {
            list.eject()
        }
    }

    if (!list.isEmpty()) {
        return false
    }

    return true
}

if (module === require.main) {
    TestRunner.parseTest(testLinkedList())
}

File: lang/ruby/test/fun/vac/data_structures/DoublyLinkedListTest.rb
require "fun/vac/data_structures/DoublyLinkedList"
require "fun/vac/util/TestRunner"

def testLinkedList()
    size = 8192

    list = DoublyLinkedList::DoublyLinkedList.new()

    (1 .. size).each do |i|
        list.append(i)
    end

    if list.size() != size
        return false
    end

    (0 ... size).each do |i|
        if list.get(i) != i + 1
            return false
        end
    end

    (0 ... size).each do |i|
        list.set(i, size - i)
    end

    (0 ... size).each do |i|
        if list.get(i) != size - i
            return false
        end
    end

    x = 0
    y = 0

    i, j = 0, size - 1
    while i < j
        x = list.get(i)
        y = list.get(j)

        list.remove(i)
        list.insert(i, y)
        list.remove(j)
        list.insert(j, x)
        i, j = i + 1, j - 1
    end

    (0 ... size).each do |i|
        if list.get(i) != i + 1
            return false
        end
    end

    i = size
    while i >= 1
        if list.back() != i
            return false
        end
        list.eject()
        i -= 1
    end

    list.isEmpty()
end

if __FILE__ == $PROGRAM_NAME
    TestRunner.parseTest(testLinkedList())
end
